### PR TITLE
Adding in GitHub vault authentication

### PR DESCRIPTION
**Pull request by apearsall**
The point of this PR is to add GitHub authentication to vault-key.  Note that this will require your Vault environment to be configured with GitHub authentication and group-mapping before it will work.  

* Adds conditional logic into config.go overriding GCP service account auth when `GITHUB_OAUTH_TOKEN` env var is set.
* Adds GitHub authentication flow into `github.go`
* Renames / moves GCP authentication into its own function `gcpSaAuth()`

### CODE REVIEW

File: pkg/vault/client.go [hunk 1]
```diff
@@ -57,14 +57,7 @@ func initClient(vc *vaultClient) error {
 	}
 
 	token, err := vc.authClient.GetVaultToken(vc)
-	if err != nil {
```
Comment (stevenaldinger) [inline pkg/vault/client.go:60]: why was this removed / .SetToken being called even if there's an error getting the token?

File: pkg/vault/client_test.go [hunk 2]
```diff
@@ -9,19 +9,35 @@ import (
...
+		serviceAccount: "none",
+		traceEnabled:   false,
+		tracePrefix:    "test",
+		vaultRole:      "read",
+	}
+
+	ghc = &config{
+		project:        "test",
+		googleAuth:     false,
+		githubAuth:     true,
```
Comment (stevenaldinger) [inline pkg/vault/client_test.go:39]: wonder if there could be an `authMethod` sort of variable here instead of multiple true/false flags. its good it works but doesn't feel like it'd scale well with more auth methods and feels wrong having to set false for things instead of just true for the one you want / it being technically possible to set multiple things to true

File: pkg/vault/vault.go [hunk 3]
```diff
@@ -22,6 +23,57 @@ type Client interface {
...
+		if err != nil {
+			return "", err
+		}
+
+		log.Debug("Successfully logged into Vault with auth/gcp/login")
+
+		return vaultResp.Auth.ClientToken, nil
+	}
+
+	return "ERROR", err
```
Comment (stevenaldinger) [inline pkg/vault/vault.go:74]: returning an empty string alongside an actual error is preferred here. you can do that like with something like this:

```
return "", errors.New("GetVaultToken: configuration error, one of [githubAuth, googleAuth] must be set to true")
```

File: pkg/vault/client.go (hunk 1 above)
Comment (apearsall) [reply pkg/vault/client.go:60]: Good call out, I was just being lazy to get tests passing.  Re-added and tests are passing now!

File: pkg/vault/vault.go (hunk 3 above)
Comment (apearsall) [reply pkg/vault/vault.go]: Good call, added and tests are passing!

File: pkg/vault/vault.go [hunk 4]
```diff
@@ -42,6 +95,10 @@ func GetSecrets(ctx context.Context, secretValues *map[string]map[string]string,
 	auth := NewAuthClient()
```
Comment (phanyzewski) [inline pkg/vault/vault.go:95]: I think here is where we should be determining the type of auth client to use.  Originally this was only setup with the gcp auth client, so that's what NewAuthClient returned, however we should rename that original auth client to NewGcpAuthClient, and create a NewGitHubAuthClient.  Both auth clients satisfy the existing auth client interface (GetVaultToken) and by adding that code here, and passing it into NewVaultClient, we don't need to modify as much of the code as we are right now.

I think it would be a bit of undoing some of the combined code you have, but it would make use of the interface and really reduce the overall complexity of adding this new auth client.

Comment (phanyzewski) [reply pkg/vault/vault.go:95]: In the end we would keep this function
```go
func (a *gcpAuthClient) GetVaultToken(vc *vaultClient) (string, error) {}
```

rename this one from 

```go
// NewAuthClient returns a new instance of an auth client
func NewAuthClient() AuthClient {
	a := &gcpAuthClient{}
	return a
}
```

to

```go
// NewAuthClient returns a new instance of an auth client
func NewGcpAuthClient() AuthClient {
	a := &gcpAuthClient{}
	return a
}
```

and add a new type and function to support the github vault client

```go
type githubAuthClient struct {
}
// NewAuthClient returns a new instance of an auth client
func NewGithubAuthClient() AuthClient {
	a := &githubAuthClient{}
	return a
}
func (a *githubAuthClient) GetVaultToken(vc *vaultClient) (string, error) {}
```

Comment (phanyzewski) [reply pkg/vault/vault.go:95]: Each auth client handles its handshake with vault in its own way, but the interface is satisfied, and the both take a vault client, and return a token or an error

**Review by phanyzewski — COMMENTED** [review]
Awesome work @apearsall, a much needed update for accessing vault with github auth.  I left a suggestion for cleaning up the client tests, and some pointers on taking advantage of the authclient interface to simplify some of the changes you've made.

**Comment by apearsall** [comment]
Leave me alone stalebot

**Comment by phanyzewski** [comment]
manually verified github auth with `osnp docker generate --dotenv`

https://github.com/teamsnap/osnp/pull/92

**Review by stevenaldinger — APPROVED** [review]
